### PR TITLE
Change websockets library

### PR DESCRIPTION
**Pull request by mikaelm12**
Still early and hacky but just want to leave this out here.Started hacking on changing the websocket library to Square's OkHttp WebSocket implementation. 
The motivation for doing this now is that we aren't keeping track of cookies which is causing problems and this library provides an easier way to do that. Still need to verify that behavior

One issue I have now is I'm a little unsure how to deal with errors during connection with these websockets. 

Issue: https://github.com/aspnet/SignalR/issues/2792
Issue: https://github.com/aspnet/SignalR/issues/3013

### CODE REVIEW

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 1]
```diff
@@ -45,6 +51,26 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
             this.logger = new NullLogger();
         }
         this.skipNegotiate = skipNegotiate;
+        this.httpClient = new OkHttpClient.Builder()
+                .cookieJar(new CookieJar() {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: @Tratcher Could you take a look at this cookie storage and see if there are any "gotchas" we should be aware of.

Comment (Tratcher) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Yeah, don't write this code. Find a real implementation.

Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: We had a nice chat about this. We'll be using the `Cookie` class from OkHttp and using `matches` and `expiresAt` to return the cookies.

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 2]
```diff
@@ -8,15 +8,17 @@
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 
-import org.java_websocket.client.WebSocketClient;
-import org.java_websocket.handshake.ServerHandshake;
+import okhttp3.*;
 
 class WebSocketTransport implements Transport {
-    private WebSocketClient webSocketClient;
+    private WebSocket newWebSocketClient;
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: `websocket` or `websocketClient`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 3]
```diff
@@ -8,15 +8,17 @@
...
+    private SignalRWebSocketListener  webSocketListener = new SignalRWebSocketListener();
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: nit: extra space

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 4]
```diff
@@ -82,36 +79,55 @@ public void onReceive(String message) throws Exception {
     @Override
     public CompletableFuture stop() {
         return CompletableFuture.runAsync(() -> {
-            webSocketClient.closeConnection(0, "HubConnection Stopped");
+            newWebSocketClient.close(0, "HubConnection stopped.");
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: `(1000, empty string)`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 5]
```diff
@@ -82,36 +79,55 @@ public void onReceive(String message) throws Exception {
...
+        public void onClosing(WebSocket webSocket, int code, String reason) {
+            logger.log(LogLevel.Information, "WebSocket connection stopping with " +
+                    "code %d and reason %d", code, reason);
+        }
+
+        @Override
+        public void onFailure(WebSocket webSocket, Throwable t, Response response) {
+            logger.log(LogLevel.Error, "Error : %d", t.getMessage());
+            // If the start future hasn't completed yet, then we need to complete it exceptionally.
+            if (!startFuture.isDone()) {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: Do the same in `onClosing`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 6]
```diff
@@ -54,6 +57,42 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
...
+                    private List<Cookie> cookieList = new ArrayList<>();
+
+                    @Override
+                    public void saveFromResponse(HttpUrl url, List<Cookie> cookies) {
+                        for (Cookie cookie : cookies) {
+                            for (Cookie innerCookie : cookieList) {
+                                if (cookie.name().equals(innerCookie.name()) && innerCookie.matches(url)) {
+                                    // We have a new cookie that matches an older one so we replace the older one.
+                                    cookieList.remove(innerCookie);
```
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: I meant `break;` not `continue;`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 7]
```diff
@@ -82,36 +79,62 @@ public void onReceive(String message) throws Exception {
...
+                onReceive(message);
+            } catch (Exception e) {
+                e.printStackTrace();
             }
-        };
+        }
+
+        @Override
+        public void onClosing(WebSocket webSocket, int code, String reason) {
+            logger.log(LogLevel.Information, "WebSocket connection stopping with " +
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: `%s` for reason

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java [hunk 8]
```diff
@@ -8,15 +8,17 @@
...
 import java.util.concurrent.CompletableFuture;
 
-import org.java_websocket.client.WebSocketClient;
-import org.java_websocket.handshake.ServerHandshake;
+import okhttp3.*;
 
 class WebSocketTransport implements Transport {
-    private WebSocketClient webSocketClient;
+    private WebSocket websocketClient;
+    private SignalRWebSocketListener webSocketListener = new SignalRWebSocketListener();
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: nit: this doesn't need to be stored

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 9]
```diff
@@ -54,6 +57,48 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
...
+                                cookieList.add(cookie);
+                            }
+                        }
+                    }
+
+                    @Override
+                    public List<Cookie> loadForRequest(HttpUrl url) {
+                        List<Cookie> matchedCookies = new ArrayList<>();
+                        List<Cookie> expiredCookies = new ArrayList<>();
+                        for (Cookie cookie: cookieList) {
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: nit: space after `cookie`

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 10]
```diff
@@ -54,6 +57,48 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
...
+                            if (cookie.expiresAt() < System.currentTimeMillis()) {
+                                expiredCookies.add(cookie);
+                            } else if (cookie.matches(url)) {
+                                matchedCookies.add(cookie);
+                            }
+                        }
+
+                        cookieList.removeAll(expiredCookies);
```
Comment (Tratcher) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Do you have to worry about any concurrency here? Will there ever be overlapping requests?

Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: Every implementation I've found looks like it assumes no overlaps, including one by one of the maintainers of the library https://gist.github.com/swankjesse/e283f2ddf264abd9a0d45bc80e7ec2d8

Comment (Tratcher) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: For your signalr scenarios is that a valid assumption? Or is concurrency managed by the calling httpclient? Java List certainly isn't thread safe.

Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: We don't guarantee requests are called in sequence. We'll give this a look and see if we need a lock.

Comment (BrennanConroy) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: From some simple testing it looks like we need a lock

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 11]
```diff
@@ -54,6 +57,61 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
...
+                                if (cookie.expiresAt() < System.currentTimeMillis()) {
+                                    expiredCookies.add(cookie);
+                                } else if (cookie.matches(url)) {
+                                    matchedCookies.add(cookie);
+                                }
+                            }
+
+                            cookieList.removeAll(expiredCookies);
+                            return matchedCookies;
+
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: nit: newline

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java [hunk 12]
```diff
@@ -54,6 +57,61 @@ public HubConnection(String url, Transport transport, Logger logger, boolean ski
...
+                                    }
+                                }
+                                if (!replacedCookie) {
+                                    cookieList.add(cookie);
+                                }
+                            }
+                        } finally {
+                            cookieLock.unlock();
+                        }
+
```
Comment (BrennanConroy) [inline clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java]: nit: newline

File: clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java (hunk 7 above)
Comment (muratg) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: Also, why is concat necessary?

Comment (mikaelm12) [reply clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/WebSocketTransport.java]: Just for line length.

**Pull request merged**